### DISCUSSION TITLE
Storing RAM (74LS189) data when the simulation is stopped?

### DISCUSSION THREAD

**Discussion by DungeonMaestro215** [hneemann/Digital#668] [Q&A]
I've been building Ben Eater's ([eater.net](url)) Simple-As-Possible breadboard computer with Digital for use in one of the University of Chapel Hill's computer science classes, and I am wondering if there is a way for the data stored in the 74LS189 (Random Access Memory) chips to carry over between stopping and starting the circuit simulation. I've attached a photo of this circuit below.

In the real world, the data in these chips is volatile, so it will go away whenever the computer is turned off; however, this can be a bit of a nuisance when students are using this model. Putting a program into this computer any more than one time starts to get tedious, so it would be nice if students could maintain their programs even when they 'power off' the model.

I'm not terribly experienced with some of the finer details of Digital, so I was wondering if there are any ideas on how this might be done. I appreciate any ideas that you may have.

![SAP](https://user-images.githubusercontent.com/44309273/109401835-ffa64a80-791e-11eb-885a-b3ca6c42ef3c.png)

**Comment by hneemann** [comment]
The easiest way will be to copy the 74189 from the library, renaming it (e.g. "my74189.dig") to avoid name collisions.
Then you can reset the "Modification locked" flag in the copy, so that the circuit becomes editable.
Now you can set the flag "program memory" in the RAM block of the copy.
After that, in the settings of the circuit of the complete computer, a hex file can be loaded into the program memory at startup. Then the students can edit the hex file with an editor and define how the program should look, which is executed at startup. 

Please let me know how it goes! :smile:

**Comment by hneemann** [comment]
I thought about it again: It's not that simple, because in the BenEater computer two 4-bit rams are connected together to one 8-bit ram. The simulator is able to distribute file contents to several rams, but only on byte level and not on bit level.
It seems to me that a "load circuit" that contains a ROM and overwrites the 74189 at the push of a button is the simplest solution.
